### ISSUE TITLE
Type inference 

### ISSUE THREAD

**Issue by olegdunkan** [Microsoft/TypeScript#16434]
```ts
class A {
    x: number;
}
class B {
    y: number;
}

function fn<T, U>(x: T & U, y: T) { }

type T = A;
type U = B;
var x: T & U;
var y: T;

fn<T,U>(x, y); // fn(x:A&B, y:A)
fn(x, y);// fn(x:A, y:A)
````
I expected that the first and second call expressions above will resolve the same call signature. 
Is it possibly to expect it to be implemented in the future?

Below example works:
```ts
function fn(x: A & B, y: A): string;
function fn(x: A, y: A):number;
function fn():string|number{
    return "";
}

var x: A & B;
var y: A;

//it works as expected
fn(x, y);//string
fn(y, y);//number

````

**Comment by mhegazy** [comment]
here is a simpler example:
```ts
function fn<T, U>(x: T & U) { }
var x: A & B;
fn(x);// T: {}, U: {}
```
The compiler tries to infer to `T & U`  from `A & B` but find no way to dive deeper into the intersection since there is no way to know how to split the input.. there is no reason why `T` should be `A` and `U` should be `B` instead of say `T` -> `{}`, `U` -> `A & B`, or  `T` -> `A&B`, `U` -> `{}`, or all other possible combinations in the middle.

If one of two types is shared in the input, i.e.:
```ts
function fn<T>(x: A & T) { }
var x: A & B;
fn(x);// T: B
```

then the compiler has a indication to how to infer to `T`.

the bottom line is type inference just a pattern matching exercise, and it needs enough similarity between source and target to make the desired inferences.

**Comment by olegdunkan** [comment]
@mhegazy thanks, I understand that it is working as intended but my question (rephrase it) was about why isn't  possibly to dive deeper? 
I expanded your case:
```ts
function fn<T, U>(x: A & U & T) { }
var x: A & B & C;
fn(x);// T: B
````
We have the same problem.
Type script now can solve equations with only the one unknown variable. 

Why don't we try to do simply assumption.
If `x` has type `A&B` we can substitute it with `T&U`.
If, for example, first argument `x` has type `A&B|C&D` and first parameter annotation is `T&U|M&N` or `T&M|T&N` or all other combinations that has the same pattern `x&x|x&x` then substitute it with corresponding types, and try next parameter.
We don't need to solve, say, linear equations.   
If we can't infer all parameter types with one pass substituting in simply manner then apply the standard inference approach.     
 
More elaborate example:
```ts
function f1<T, U, P>(x: T & (U | P), y: U | P, z: P);
function f2<T, U, P>(z: P, y: U | P, x: T & (U | P));

var x: A & (B | C);
var y: B | C;
var z: C;

f1(x, y, z); //pass 
f2(z, y, x); //pass 
````
